### PR TITLE
[class] relax assumption about origin of a class (can come from an dynamic image)

### PR DESCRIPTION
**Pull request by lewurm**
check got introduced with ffc3aaa95c45fe6888f64572df35e3fe7ba4f8ac (in most places).

fixes https://bugzilla.xamarin.com/show_bug.cgi?id=41130

### CODE REVIEW

**Comment by lewurm** [comment]
not sure if it's the right fix, please review and comment 🙂

File: mono/metadata/class-inlines.h [hunk 1]
```diff
@@ -88,9 +88,9 @@ mono_class_is_public (MonoClass *klass)
 }
 
 static inline gboolean
-mono_class_has_static_metadata (MonoClass *klass)
+mono_class_has_metadata (MonoClass *klass)
 {
-	return klass->type_token && !klass->image->dynamic && !mono_class_is_ginst (klass);
+	return klass->type_token && !mono_class_is_ginst (klass);
```
Comment (lewurm) [inline mono/metadata/class-inlines.h]: this is the actual change

**Comment by lambdageek** [comment]
Quick update.  I took the test case `bug-41130.cs` and changed it to use reflection to create a `x = new Foo<int>(); x.value = 42` instance and then to call `Bar.Get<int> (x)`.  That works as expected (returns 42).

So the problem is just the `MetadataToken` property of `Foo<A>.value` where `A` is the generic parameter of the generic method `static A Bar.Get<A> (Foo<A>)`.

**Comment by lambdageek** [comment]
So the problem is in `mono_reflection_get_token_checked`.  It's job is: given a managed reflection object, return the `MetadataToken` apropriate for that object.

In this case it's called on a `System.Reflection.MonoField` instance - but it so happens that the `SR.MonoField` is for a field of a class that was created using reflection.  Unfortunately, the only things that a `SR.MonoField` has access to are the underlying `MonoClassField*`.

`MonoClassField*` does not store a metadata token.  This makes sense for classes that come from ordinary `MonoImage`s: the fields of a class are assigned consecutive tokens starting from some initial index - that's what `mono_class_get_field_token` basically computes: get the first metadata token from tables and compute the `MonoClassField*` offset from `&klass.fields[0]`.

The problem (as I mentioned [above](https://github.com/mono/mono/pull/6035#pullrequestreview-77464082)) is that for dynamic classes the offset of a `MonoClassField*` relative to `&klass.fields[0]` has no relation to its token.

Note that this is not a problem for methods because `MonoMethod*` has a token field.

I think there are two possible ways to go:
1. Provide a mechanism to get from a `SR.MonoField` in a dynamic image back to its `SRE.FieldBuilder` (which stores the token that was assigned at definition time).  One way to do this is to have a map in `SRE.ModuleBuilder` that maps each `SRe.FieldBuilder` to a baked `SR.MonoField`. Advantage: no cost for non-SRE cases. Disadvantage: I'm not sure if the map can be done lazily.
2. or just add a token field to `MonoClassField`.  Advantage: no maps; lookup is fast and simple. Disadvantage: all classes get bigger.
3. or maybe add a member to `SR.MonoField` (either a `SRE.FieldBuilder wasFieldBiulder` or a `int sreToken`) if the member is set then use it to return `MetadataToken`; assume it wasn't an SRE case and dive down to the runtime  as we do now.

**Comment by lambdageek** [comment]
Related: https://bugzilla.xamarin.com/show_bug.cgi?id=33208

Overall the constraints are:
1. from managed: each `SR.Module` and `MetadataToken` ought to map uniquely to a `SR.MonoField`
2. in the runtime: we really want consecutive `MonoClassField*`s of a `MonoClass` to have consecutive tokens.

I tried to get something that accomplishes both of those things in https://github.com/mono/mono/pull/3908 but I never got it reviewed or merged. It might be worth reviving.  But I don't think I properly considered an example like the one in 41130.

**Pull request closed without merging**